File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/python/dp_algorithms/egg_drop.py
from __future__ import annotations

from functools import lru_cache

from dp_algorithms.common import DPResult


def memoized(eggs: int, floors: int) -> DPResult:
    @lru_cache(maxsize=None)
    def solve(e: int, f: int) -> int:
        if f <= 1 or e == 1:
            return f
        return 1 + min(max(solve(e - 1, x - 1), solve(e, f - x)) for x in range(1, f + 1))

    return DPResult("Egg Drop", "memoization", solve(eggs, floors))


def tabulated(eggs: int, floors: int) -> DPResult:
    dp = [[0] * (floors + 1) for _ in range(eggs + 1)]
    for e in range(1, eggs + 1):
        dp[e][1:] = range(1, floors + 1)
    for e in range(2, eggs + 1):
        for f in range(2, floors + 1):
            dp[e][f] = 1 + min(max(dp[e - 1][x - 1], dp[e][f - x]) for x in range(1, f + 1))
    return DPResult("Egg Drop", "tabulation", dp[eggs][floors])


def space_optimized(eggs: int, floors: int) -> DPResult:
    moves = 0
    reachable = [0] * (eggs + 1)
    while reachable[eggs] < floors:
        moves += 1
        for e in range(eggs, 0, -1):
            reachable[e] = reachable[e] + reachable[e - 1] + 1
    return DPResult("Egg Drop", "moves/space optimized", moves, {"reachable_floors": reachable[eggs]})

File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/java/src/main/java/com/example/dp/EggDrop.java
package com.example.dp;

import java.util.HashMap;
import java.util.Map;

public final class EggDrop {
    private EggDrop() {
    }

    public static Result memoized(int eggs, int floors) {
        return new Result("Egg Drop", "memoization", solve(eggs, floors, new HashMap<>()), Map.of());
    }

    private static int solve(int eggs, int floors, Map<String, Integer> memo) {
        if (floors <= 1 || eggs == 1) {
            return floors;
        }
        String key = eggs + ":" + floors;
        if (memo.containsKey(key)) {
            return memo.get(key);
        }
        int best = Integer.MAX_VALUE;
        for (int floor = 1; floor <= floors; floor++) {
            best = Math.min(best, 1 + Math.max(solve(eggs - 1, floor - 1, memo), solve(eggs, floors - floor, memo)));
        }
        memo.put(key, best);
        return best;
    }

    public static Result tabulated(int eggs, int floors) {
        int[][] dp = new int[eggs + 1][floors + 1];
        for (int egg = 1; egg <= eggs; egg++) {
            for (int floor = 1; floor <= floors; floor++) {
                dp[egg][floor] = floor;
            }
        }
        for (int egg = 2; egg <= eggs; egg++) {
            for (int floor = 2; floor <= floors; floor++) {
                for (int drop = 1; drop <= floor; drop++) {
                    dp[egg][floor] = Math.min(dp[egg][floor], 1 + Math.max(dp[egg - 1][drop - 1], dp[egg][floor - drop]));
                }
            }
        }
        return new Result("Egg Drop", "tabulation", dp[eggs][floors], Map.of());
    }

    public static Result spaceOptimized(int eggs, int floors) {
        int moves = 0;
        int[] reachable = new int[eggs + 1];
        while (reachable[eggs] < floors) {
            moves++;
            for (int egg = eggs; egg >= 1; egg--) {
                reachable[egg] = reachable[egg] + reachable[egg - 1] + 1;
            }
        }
        return new Result("Egg Drop", "moves/space optimized", moves, Map.of("reachable_floors", reachable[eggs]));
    }
}
